Sources/TDS/Message/Login7Message.swift

```diff
@@ -23,18 +58,80 @@ extension TDSMessages {
         var language: String
         var database: String
         var sspiData: String
-        
+        var atchDBFile: String = "" // TODO: What is this?
+        var changePassword: String = ""
         
         public func serialize(into buffer: inout ByteBuffer) throws {
-            // Packet Header: 0x00 - 0x08 (8 bytes)
+            // Each basic field needs to serialize the length & offset
+            let basicFields = [
+                hostname,
+                username,
+                password,
+                appName,
+                serverName,
+                "", // unused field
+                clientInterfaceName,
+                language,
+                database
+            ]
+            
+            // ClientID serializes inbetween `basicFields` and `extendedFields`
+            let clientId = [UInt8](repeating: 0, count: 6)
+            
+            // Each extended field needs to serialize the length & offset
+            let extendedFields = [
+                sspiData,
+                atchDBFile,
+                changePassword
+            ]
+            
+            let sspiLong: UInt32 = 0
+            
+            // Stores the position and skips an UInt32 so the length can be added later
+            let login7HeaderPosition = buffer.writerIndex
+            buffer.moveWriterIndex(forwardBy: 4)
+            
             buffer.writeBytes([
-                Login7Message.headerType.value,         // Type
-                0x01,                                   // Status
-                0x00, PreloginPacket.messageLength,    // Length
-                0x00, 0x00,                             // SPID
-                0x00,                                   // PacketID (Unused)
-                0x00                                    // Window (Unused)
+                0x02, 0x00, 0x09, 0x72, // TDS version
+                0x00, 0x10, 0x00, 0x00, // Packet length negotiation
+                0x00, 0x00, 0x00, 0x01, // Client version
             ])
+            
+            buffer.writeInteger(Self.clientPID)
+            buffer.writeInteger(0 as UInt32) // Connection ID
+            buffer.writeInteger(0xE3 as UInt8) // Flags1
+            buffer.writeInteger(0x03 as UInt8) // Flags2
+            buffer.writeInteger(0 as UInt8) // Flags
+            buffer.writeInteger(0 as UInt8) // Flags3
+            buffer.writeInteger(0 as UInt32) // Timezone
+            buffer.writeBytes([0x09, 0x04, 0x00, 0x00]) // ClientLCID
+            
+            var offsetLengthsPosition = buffer.writerIndex
+            buffer.moveWriterIndex(forwardBy: basicFields.count * 4)
+            buffer.writeBytes(clientId)
+            buffer.moveWriterIndex(forwardBy: extendedFields.count * 4)
+            
+            func writeField(_ string: String) {
+                let utf16 = string.utf16
```

Comment:
why it's utf16 and not UTF-8?

Reply:
It's a bit experimental still, but as far as I could tell from the protocol spec that's how this packet is formed.

This is the example packet:

```
10 01 00 90 00 00 01 00 88 00 00 00 02 00 09 72 00 10 00 00 00 00 00 07 00 01 00 00 00 00 00 00 E0 03 00 00 E0 01 00 00 09 04 00 00 5E 00 08 00 6E 00 02 00 72 00 00 00 72 00 07 00 80 00 00 00 80 00 00 00 80 00 04 00 88 00 00 00 88 00 00 00 00 50 8B E2 B7 8F 88 00 00 00 88 00 00 00 88 00 00 00 00 00 00 00 73 00 6B 00 6F 00 73 00 74 00 6F 00 76 00 31 00 73 00 61 00 4F 00 53 00 51 00 4C 00 2D 00 33 00 32 00 4F 00 44 00 42 00 43 00
```

Many `0x00` bytes refer to a smaller integer for offset or length, but at the end of the packet it refers to textual values. Specifically this bit:

```
73 00 6B 00 6F 00 73 00 74 00 6F 00 76 00 31 00 73 00 61 00 4F 00 53 00 51 00 4C 00 2D 00 33 00 32 00 4F 00 44 00 42 00 43 00
```